File: sdk/src/main/java/com/cxense/cxensesdk/AdvertisingIdProvider.java
package com.cxense.cxensesdk;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.gms.ads.identifier.AdvertisingIdClient;
import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
import com.google.android.gms.common.GooglePlayServicesRepairableException;

import java.io.IOException;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * @author Dmitriy Konopelkin (dmitry.konopelkin@cxense.com) on (2018-09-17).
 */
class AdvertisingIdProvider {
    private static final String TAG = AdvertisingIdProvider.class.getSimpleName();
    private static final long DELAY = 300;
    private final Context context;
    private final ScheduledExecutorService executor;
    private final Runnable getAdvertisingInfoTask;

    private AdvertisingIdClient.Info advertisingInfo;

    AdvertisingIdProvider(@NonNull Context context, @NonNull ScheduledExecutorService executor) {
        this.context = context;
        this.executor = executor;
        getAdvertisingInfoTask = () -> {
            try {
                advertisingInfo = AdvertisingIdClient.getAdvertisingIdInfo(this.context);
            } catch (IOException | GooglePlayServicesRepairableException e) {
                Log.e(TAG, e.getMessage(), e);
                initAdvertisingIdTask();
            } catch (GooglePlayServicesNotAvailableException e) {
                Log.e(TAG, e.getMessage(), e);
            }
        };
        initAdvertisingIdTask();
    }

    private void initAdvertisingIdTask() {
        executor.schedule(getAdvertisingInfoTask, DELAY, TimeUnit.MILLISECONDS);
    }

    @Nullable
    String getDefaultUserId() {
        return advertisingInfo != null ? advertisingInfo.getId() : null;
    }

    boolean isLimitAdTrackingEnabled() {
        return advertisingInfo != null && advertisingInfo.isLimitAdTrackingEnabled();
    }
}

File: sdk/src/main/java/io/piano/android/cxense/AdvertisingIdProvider.kt
package io.piano.android.cxense

import android.content.Context
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.google.android.gms.common.GooglePlayServicesRepairableException
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * Provides AAID
 */
class AdvertisingIdProvider(
    private val context: Context,
    private val executor: ScheduledExecutorService,
) {
    private var advertisingInfo: AdvertisingIdClient.Info? = null

    private val getAdvertisingInfoTask: Runnable = Runnable {
        try {
            advertisingInfo = AdvertisingIdClient.getAdvertisingIdInfo(context)
        } catch (e: Exception) {
            Timber.e(e)
            if (e is IOException || e is GooglePlayServicesRepairableException) {
                initAdvertisingIdTask()
            }
        }
    }

    private fun initAdvertisingIdTask() = executor.schedule(getAdvertisingInfoTask, DELAY, TimeUnit.MILLISECONDS)

    val defaultUserId: String?
        get() = advertisingInfo?.id

    val limitAdTrackingEnabled: Boolean
        get() = advertisingInfo?.isLimitAdTrackingEnabled ?: false

    init {
        initAdvertisingIdTask()
    }

    companion object {
        private const val DELAY: Long = 300
    }
}
